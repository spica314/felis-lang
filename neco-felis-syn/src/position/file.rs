use crate::{Item, Parse, ParseError, Phase, PhaseParse, token::Token};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct File<P: Phase> {
    pub items: Vec<Item<P>>,
    pub ext: P::FileExt,
}

impl<P: Phase> File<P> {
    pub fn items(&self) -> &[Item<P>] {
        &self.items
    }
}

impl Parse for File<PhaseParse> {
    fn parse(tokens: &[Token], i: &mut usize) -> Result<Option<Self>, ParseError> {
        let mut k = *i;

        let mut items = vec![];
        while let Some(item) = Item::parse(tokens, &mut k)? {
            items.push(item);
        }

        let file = File { items, ext: () };

        *i = k;
        Ok(Some(file))
    }
}

#[cfg(test)]
mod test {
    use crate::{FileIdGenerator, token::Token};

    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_parse_inductive_eq() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/inductive_eq.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }

    #[test]
    fn test_parse_inductive_nat() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/inductive_nat.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }

    #[test]
    fn test_parse_eq_and_nat() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/eq_and_nat.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }

    #[test]
    fn test_parse_exit_42() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/exit_42.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }

    #[test]
    fn test_parse_let_mut() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/let_mut.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let result = File::parse(&tokens, &mut i);

        match result {
            Ok(Some(file)) => {
                println!("Parsed successfully up to token {i} of {}", tokens.len());
                if i < tokens.len() {
                    println!("Remaining tokens: {:?}", &tokens[i..]);
                }
                assert_debug_snapshot!(file);
                assert_eq!(i, tokens.len());
            }
            Ok(None) => {
                println!("Parsing returned None at token {i} of {}", tokens.len());
                println!(
                    "Tokens around position: {:?}",
                    &tokens[i.saturating_sub(3)..tokens.len().min(i + 3)]
                );
                panic!("Parsing returned None");
            }
            Err(e) => {
                println!("Parse error at token {i} of {}: {:?}", tokens.len(), e);
                println!(
                    "Tokens around position: {:?}",
                    &tokens[i.saturating_sub(3)..tokens.len().min(i + 3)]
                );
                panic!("Parse error: {e:?}");
            }
        }
    }

    #[test]
    fn test_parse_array() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/array.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let result = File::parse(&tokens, &mut i);

        match result {
            Ok(Some(file)) => {
                println!("Parsed successfully up to token {i} of {}", tokens.len());
                if i < tokens.len() {
                    println!("Remaining tokens: {:?}", &tokens[i..]);
                }
                assert_debug_snapshot!(file);

                // Complete file parsing including #proc block is now working
                assert_eq!(i, tokens.len()); // Should have parsed all tokens

                // Verify the array was parsed correctly - it should be the 6th item (0-indexed)
                assert_eq!(file.items.len(), 7);
                match &file.items[5] {
                    crate::Item::Array(array) => {
                        assert_eq!(array.name().s(), "Points");
                        assert_eq!(array.fields().len(), 2);
                    }
                    _ => panic!("Expected 5th item to be an array"),
                }
            }
            Ok(None) => {
                println!("Parsing returned None at token {i} of {}", tokens.len());
                println!(
                    "Tokens around position: {:?}",
                    &tokens[i.saturating_sub(3)..tokens.len().min(i + 3)]
                );
                panic!("Parsing returned None");
            }
            Err(e) => {
                println!("Parse error at token {i} of {}: {:?}", tokens.len(), e);
                println!(
                    "Tokens around position: {:?}",
                    &tokens[i.saturating_sub(3)..tokens.len().min(i + 3)]
                );
                panic!("Parse error: {e:?}");
            }
        }
    }

    #[test]
    fn test_parse_struct_1() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/struct_1.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());

        // Verify the struct was parsed correctly
        // Look for the struct item in the parsed items
        let mut found_struct = false;
        for item in &file.items {
            if let crate::Item::Struct(struct_item) = item {
                assert_eq!(struct_item.name().s(), "Vec3");
                assert_eq!(struct_item.fields().len(), 3);
                assert_eq!(struct_item.fields()[0].name.s(), "x");
                assert_eq!(struct_item.fields()[1].name.s(), "y");
                assert_eq!(struct_item.fields()[2].name.s(), "z");
                found_struct = true;
                break;
            }
        }
        assert!(found_struct, "Vec3 struct not found in parsed items");
    }

    #[test]
    fn test_parse_array_len() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/array_len.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        println!("Total tokens: {}", tokens.len());
        for (i, token) in tokens.iter().enumerate() {
            println!("{i}: {token:?}");
        }

        let mut i = 0;
        let result = File::parse(&tokens, &mut i);

        // Debug: Print status
        if i >= 44 {
            println!("Stopped at token {i}, trying to parse remaining tokens as ItemProc");
        }

        match result {
            Ok(Some(file)) => {
                println!("Parsed successfully up to token {i} of {}", tokens.len());
                if i < tokens.len() {
                    println!("Remaining tokens: {:?}", &tokens[i..]);
                }
                assert_debug_snapshot!(file);

                // Complete file parsing
                assert_eq!(i, tokens.len()); // Should have parsed all tokens
            }
            Ok(None) => {
                println!("Parsing returned None at token {i} of {}", tokens.len());
                println!(
                    "Tokens around position: {:?}",
                    &tokens[i.saturating_sub(3)..tokens.len().min(i + 3)]
                );
                panic!("Parsing returned None");
            }
            Err(e) => {
                println!("Parse error at token {i} of {}: {:?}", tokens.len(), e);
                println!(
                    "Tokens around position: {:?}",
                    &tokens[i.saturating_sub(3)..tokens.len().min(i + 3)]
                );
                panic!("Parse error: {e:?}");
            }
        }
    }
}
