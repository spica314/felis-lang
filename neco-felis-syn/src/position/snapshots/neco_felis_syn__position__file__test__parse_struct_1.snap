---
source: neco-felis-syn/src/position/file.rs
expression: file
---
File {
    items: [
        Entrypoint(
            ItemEntrypoint {
                keyword_entrypoint: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 1,
                    },
                    s: "entrypoint",
                },
                name: TokenVariable {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 13,
                    },
                    s: "main",
                },
                semicolon: TokenSemicolon {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 17,
                    },
                },
                ext: (),
            },
        ),
        UseBuiltin(
            ItemUseBuiltin {
                keyword_use_builtin: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 3,
                        column: 1,
                    },
                    s: "use_builtin",
                },
                builtin_name: TokenString {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 3,
                        column: 14,
                    },
                    s: "syscall",
                },
                keyword_as: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 3,
                        column: 24,
                    },
                    s: "as",
                },
                name: TokenVariable {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 3,
                        column: 28,
                    },
                    s: "__syscall",
                },
                semicolon: TokenSemicolon {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 3,
                        column: 37,
                    },
                },
                ext: (),
            },
        ),
        UseBuiltin(
            ItemUseBuiltin {
                keyword_use_builtin: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 4,
                        column: 1,
                    },
                    s: "use_builtin",
                },
                builtin_name: TokenString {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 4,
                        column: 14,
                    },
                    s: "u64_add",
                },
                keyword_as: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 4,
                        column: 24,
                    },
                    s: "as",
                },
                name: TokenVariable {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 4,
                        column: 28,
                    },
                    s: "__u64_add",
                },
                semicolon: TokenSemicolon {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 4,
                        column: 37,
                    },
                },
                ext: (),
            },
        ),
        UseBuiltin(
            ItemUseBuiltin {
                keyword_use_builtin: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 5,
                        column: 1,
                    },
                    s: "use_builtin",
                },
                builtin_name: TokenString {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 5,
                        column: 14,
                    },
                    s: "u64",
                },
                keyword_as: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 5,
                        column: 20,
                    },
                    s: "as",
                },
                name: TokenVariable {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 5,
                        column: 24,
                    },
                    s: "__u64",
                },
                semicolon: TokenSemicolon {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 5,
                        column: 29,
                    },
                },
                ext: (),
            },
        ),
        Struct(
            ItemStruct {
                keyword_struct: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 7,
                        column: 1,
                    },
                    s: "struct",
                },
                name: TokenVariable {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 7,
                        column: 9,
                    },
                    s: "Vec3",
                },
                brace_l: TokenBraceL {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 7,
                        column: 14,
                    },
                },
                fields: [
                    ItemStructField {
                        name: TokenVariable {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 8,
                                column: 5,
                            },
                            s: "x",
                        },
                        colon: TokenColon {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 8,
                                column: 6,
                            },
                        },
                        ty: Variable(
                            TermVariable {
                                variable: TokenVariable {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 8,
                                        column: 8,
                                    },
                                    s: "__u64",
                                },
                                ext: (),
                            },
                        ),
                        comma: Some(
                            TokenComma {
                                pos: Pos {
                                    file_id: FileId(
                                        0,
                                    ),
                                    line: 8,
                                    column: 13,
                                },
                            },
                        ),
                    },
                    ItemStructField {
                        name: TokenVariable {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 9,
                                column: 5,
                            },
                            s: "y",
                        },
                        colon: TokenColon {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 9,
                                column: 6,
                            },
                        },
                        ty: Variable(
                            TermVariable {
                                variable: TokenVariable {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 9,
                                        column: 8,
                                    },
                                    s: "__u64",
                                },
                                ext: (),
                            },
                        ),
                        comma: Some(
                            TokenComma {
                                pos: Pos {
                                    file_id: FileId(
                                        0,
                                    ),
                                    line: 9,
                                    column: 13,
                                },
                            },
                        ),
                    },
                    ItemStructField {
                        name: TokenVariable {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 10,
                                column: 5,
                            },
                            s: "z",
                        },
                        colon: TokenColon {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 10,
                                column: 6,
                            },
                        },
                        ty: Variable(
                            TermVariable {
                                variable: TokenVariable {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 10,
                                        column: 8,
                                    },
                                    s: "__u64",
                                },
                                ext: (),
                            },
                        ),
                        comma: Some(
                            TokenComma {
                                pos: Pos {
                                    file_id: FileId(
                                        0,
                                    ),
                                    line: 10,
                                    column: 13,
                                },
                            },
                        ),
                    },
                ],
                brace_r: TokenBraceR {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 11,
                        column: 1,
                    },
                },
                ext: (),
            },
        ),
        Proc(
            ItemProc {
                ptx_modifier: None,
                keyword_proc: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 13,
                        column: 1,
                    },
                    s: "proc",
                },
                name: TokenVariable {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 13,
                        column: 7,
                    },
                    s: "main",
                },
                colon: TokenColon {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 13,
                        column: 12,
                    },
                },
                ty: ArrowNodep(
                    TermArrowNodep {
                        from: Unit(
                            TermUnit {
                                paren_l: TokenParenL {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 13,
                                        column: 14,
                                    },
                                },
                                paren_r: TokenParenR {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 13,
                                        column: 15,
                                    },
                                },
                                ext: (),
                            },
                        ),
                        arrow: TokenOperator {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 13,
                                column: 17,
                            },
                            s: "->",
                            after_whitespace: true,
                        },
                        to: Unit(
                            TermUnit {
                                paren_l: TokenParenL {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 13,
                                        column: 20,
                                    },
                                },
                                paren_r: TokenParenR {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 13,
                                        column: 21,
                                    },
                                },
                                ext: (),
                            },
                        ),
                        ext: (),
                    },
                ),
                proc_block: ItemProcBlock {
                    brace_l: TokenBraceL {
                        pos: Pos {
                            file_id: FileId(
                                0,
                            ),
                            line: 13,
                            column: 23,
                        },
                    },
                    statements: Then(
                        StatementsThen {
                            head: Let(
                                StatementLet {
                                    let_keyword: TokenKeyword {
                                        pos: Pos {
                                            file_id: FileId(
                                                0,
                                            ),
                                            line: 14,
                                            column: 5,
                                        },
                                        s: "let",
                                    },
                                    variable: TokenVariable {
                                        pos: Pos {
                                            file_id: FileId(
                                                0,
                                            ),
                                            line: 14,
                                            column: 10,
                                        },
                                        s: "syscall_id",
                                    },
                                    equals: TokenOperator {
                                        pos: Pos {
                                            file_id: FileId(
                                                0,
                                            ),
                                            line: 14,
                                            column: 21,
                                        },
                                        s: "=",
                                        after_whitespace: true,
                                    },
                                    value: Number(
                                        ProcTermNumber {
                                            number: TokenNumber {
                                                pos: Pos {
                                                    file_id: FileId(
                                                        0,
                                                    ),
                                                    line: 14,
                                                    column: 23,
                                                },
                                                s: "231u64",
                                            },
                                            ext: (),
                                        },
                                    ),
                                    ext: (),
                                },
                            ),
                            semicolon: TokenSemicolon {
                                pos: Pos {
                                    file_id: FileId(
                                        0,
                                    ),
                                    line: 14,
                                    column: 29,
                                },
                            },
                            tail: Then(
                                StatementsThen {
                                    head: Let(
                                        StatementLet {
                                            let_keyword: TokenKeyword {
                                                pos: Pos {
                                                    file_id: FileId(
                                                        0,
                                                    ),
                                                    line: 15,
                                                    column: 5,
                                                },
                                                s: "let",
                                            },
                                            variable: TokenVariable {
                                                pos: Pos {
                                                    file_id: FileId(
                                                        0,
                                                    ),
                                                    line: 15,
                                                    column: 10,
                                                },
                                                s: "error_code",
                                            },
                                            equals: TokenOperator {
                                                pos: Pos {
                                                    file_id: FileId(
                                                        0,
                                                    ),
                                                    line: 15,
                                                    column: 21,
                                                },
                                                s: "=",
                                                after_whitespace: true,
                                            },
                                            value: Number(
                                                ProcTermNumber {
                                                    number: TokenNumber {
                                                        pos: Pos {
                                                            file_id: FileId(
                                                                0,
                                                            ),
                                                            line: 15,
                                                            column: 23,
                                                        },
                                                        s: "42",
                                                    },
                                                    ext: (),
                                                },
                                            ),
                                            ext: (),
                                        },
                                    ),
                                    semicolon: TokenSemicolon {
                                        pos: Pos {
                                            file_id: FileId(
                                                0,
                                            ),
                                            line: 15,
                                            column: 25,
                                        },
                                    },
                                    tail: Then(
                                        StatementsThen {
                                            head: Expr(
                                                Apply(
                                                    ProcTermApply {
                                                        f: Variable(
                                                            ProcTermVariable {
                                                                variable: TokenVariable {
                                                                    pos: Pos {
                                                                        file_id: FileId(
                                                                            0,
                                                                        ),
                                                                        line: 16,
                                                                        column: 5,
                                                                    },
                                                                    s: "__syscall",
                                                                },
                                                                ext: (),
                                                            },
                                                        ),
                                                        args: [
                                                            Variable(
                                                                ProcTermVariable {
                                                                    variable: TokenVariable {
                                                                        pos: Pos {
                                                                            file_id: FileId(
                                                                                0,
                                                                            ),
                                                                            line: 16,
                                                                            column: 15,
                                                                        },
                                                                        s: "syscall_id",
                                                                    },
                                                                    ext: (),
                                                                },
                                                            ),
                                                            Variable(
                                                                ProcTermVariable {
                                                                    variable: TokenVariable {
                                                                        pos: Pos {
                                                                            file_id: FileId(
                                                                                0,
                                                                            ),
                                                                            line: 16,
                                                                            column: 26,
                                                                        },
                                                                        s: "error_code",
                                                                    },
                                                                    ext: (),
                                                                },
                                                            ),
                                                            Number(
                                                                ProcTermNumber {
                                                                    number: TokenNumber {
                                                                        pos: Pos {
                                                                            file_id: FileId(
                                                                                0,
                                                                            ),
                                                                            line: 16,
                                                                            column: 37,
                                                                        },
                                                                        s: "0",
                                                                    },
                                                                    ext: (),
                                                                },
                                                            ),
                                                            Number(
                                                                ProcTermNumber {
                                                                    number: TokenNumber {
                                                                        pos: Pos {
                                                                            file_id: FileId(
                                                                                0,
                                                                            ),
                                                                            line: 16,
                                                                            column: 39,
                                                                        },
                                                                        s: "0",
                                                                    },
                                                                    ext: (),
                                                                },
                                                            ),
                                                            Number(
                                                                ProcTermNumber {
                                                                    number: TokenNumber {
                                                                        pos: Pos {
                                                                            file_id: FileId(
                                                                                0,
                                                                            ),
                                                                            line: 16,
                                                                            column: 41,
                                                                        },
                                                                        s: "0",
                                                                    },
                                                                    ext: (),
                                                                },
                                                            ),
                                                            Number(
                                                                ProcTermNumber {
                                                                    number: TokenNumber {
                                                                        pos: Pos {
                                                                            file_id: FileId(
                                                                                0,
                                                                            ),
                                                                            line: 16,
                                                                            column: 43,
                                                                        },
                                                                        s: "0",
                                                                    },
                                                                    ext: (),
                                                                },
                                                            ),
                                                        ],
                                                        ext: (),
                                                    },
                                                ),
                                            ),
                                            semicolon: TokenSemicolon {
                                                pos: Pos {
                                                    file_id: FileId(
                                                        0,
                                                    ),
                                                    line: 16,
                                                    column: 44,
                                                },
                                            },
                                            tail: Nil,
                                            ext: (),
                                        },
                                    ),
                                    ext: (),
                                },
                            ),
                            ext: (),
                        },
                    ),
                    brace_r: TokenBraceR {
                        pos: Pos {
                            file_id: FileId(
                                0,
                            ),
                            line: 17,
                            column: 1,
                        },
                    },
                    ext: (),
                },
                ext: (),
            },
        ),
    ],
    ext: (),
}
